src/binary/cursor.rs

```diff
@@ -22,18 +22,114 @@ use crate::{
 };
 use std::io;
 
+use std::ops::Range;
+
 /// Information about the value over which the Cursor is currently positioned.
 #[derive(Clone, Debug)]
 struct EncodedValue {
     ion_type: IonType,
     header: Header,
     is_null: bool,
     index_at_depth: usize,
-    length_in_bytes: usize,
-    last_byte: usize,
     field_id: Option<SymbolId>,
     annotations: Vec<SymbolId>,
     parent_index: Option<usize>,
+
+    // Each encoded value has up to four components, appearing
+    // in the following order:
+    //
+    // [ field_id? | annotations? | header | value ]
+    //
+    // The following fields store their respective byte offsets
+    // from the beginning of the data source.
+
+    // Only valid if `field_id` is defined.
+    field_id_offset: usize,
+    // Only valid if `annotations` is not empty.
+    annotations_offset: usize,
+    // Type descriptor byte location.
+    header_offset: usize,
+    // The first byte after the type descriptor and 'length' field.
+    // Only valid if `value_length` is non-zero.
+    value_offset: usize,
+    // The number of bytes used to encode the value, not including
+    // the header byte or length fields.
+    value_length: usize,
+
+    // The first offset beyond the end of this value, calculated as:
+    //     value_offset + value_length
+    // Stored here to avoid having to recalculate it.
+    value_end: usize,
```

Comment:
`value_end` (an exclusive range end) replaced the `last_byte` field, which was an inclusive range end. This allowed me to simplify some bounds checking.